mlos_bench/mlos_bench/run.py

```diff
@@ -154,29 +123,96 @@ def _optimize(*,
     return (best_score, best_config)
 
 
-def _run(env: Environment, opt: Optimizer, trial: Storage.Trial, global_config: Dict[str, Any]) -> None:
+def _load_config(exp: Storage.Experiment, env_context: Environment,
+                 config_id: int) -> TunableGroups:
+    """
+    Load the existing tunable configuration from the storage.
+    """
+    tunable_values = exp.load_tunable_config(config_id)
+    tunables = env_context.tunable_params.assign(tunable_values)
+    _LOG.info("Load config from storage: %d", config_id)
+    if _LOG.isEnabledFor(logging.DEBUG):
+        _LOG.debug("Config %d ::\n%s",
+                   config_id, json.dumps(tunable_values, indent=2))
+    return tunables
+
+
+def _scheduler(exp: Storage.Experiment, env_context: Environment,
+               global_config: Dict[str, Any], running: bool = False) -> None:
+    """
+    Scheduler part of the loop. Check for pending trials in the queue and run them.
+    """
+    for trial in exp.pending_trials(datetime.utcnow(), running=running):
+        _run_trial(env_context, trial, global_config)
+
+
+def _optimizer(exp: Storage.Experiment, opt_context: Optimizer,
+               last_trial_id: int = -1, trial_config_repeat_count: int = 1,
+               is_warm_up: bool = False) -> int:
+    """
+    Optimizer part of the loop. Load the results of the executed trials
+    into the optimizer, suggest new configurations, and add them to the queue.
+    Return the last trial ID processed by the optimizer.
+    """
+    (trial_ids, configs, scores, status) = exp.load(last_trial_id)
+    opt_context.bulk_register(configs, scores, status, is_warm_up)
```

Comment:
Another possibility here:
Perhaps we don't just bulk register, but entirely retrain an optimizer with the latest data on each round.

For instance, not all the optimizers we have support an internal notion of "noise", so we can't always just feed the `[(config_a, score_1), (config_a, score_2,), ...]` to the optimizer, we might actually need to feed `[(config_a, aggregate_score)]` in, which we won't have until after the fact, which means *replacing* prior Optimizer models, not just augmenting them.

Reply:
I think this is more of an optimizer policy than the queue one. Regardless of the optimizer, we warm it up with the prior data, and then feed more results as we get them. It is up to the optimizer to decide whether and how to retain this data and produce the results. In fact, BO uses *all* data points every time it produces a suggestion. Let's keep it simple in the optimization loop and add that retraining logic to the optimizer when we need it (and, again, not all optimizers will need it anyway)

Reply:
In principle I agree, but BO doesn't always do that - it depends on the implementation of the library.  If you look at the output for some of these libraries, they claim to *replace* the `(config_a, score_x)` value instead of augmenting the dataset.

That's what makes me think we might want to be able to manage that for ourselves on occasion, but I think you're right - it's not a Scheduler problem.  Rather it's an interaction between the Optimizer and the Storage class (assuming we don't want to keep all historical records in memory all the time).